snapcraft/storeapi/__init__.py

```diff
@@ -387,6 +399,63 @@ def snap_release(self, snap_name, revision, channels):
 
         return response_json
 
+    def generate_snap_build(self, account_info, snap_name,
+                            snap_filename, grade, key_name):
+        assertion_file = sub(r"(.*).snap$", r"\1.snap-build", snap_filename)
+        if os.path.isfile(assertion_file):
+            logger.debug('Skipping snap-build generation, it already exists')
+            return
+
+        try:
+            authority_id = account_info['account_id']
+            snaps = account_info['snaps'][constants.DEFAULT_SERIES]
+            snap_id = snaps[snap_name]['snap-id']
+        except KeyError as err:
+            msg = 'Could not get all account information needed'
+            raise snapcraft.internal.meta.CommandError(msg)
+
+        with open(assertion_file, 'w+') as outfile:
+            try:
+                # snap-digest is calculated by snap sign-build itself
+                # snap-size is calculated by snap sign-build itself
+                # timestamp is taken by snap sign-build itself
+                cmd = ['snap', 'sign-build',
+                       '--developer-id=' + authority_id,
+                       '--snap-id=' + snap_id,
+                       '--grade=' + grade,
+                       '-k', key_name,
+                       snap_filename]
+                logger.debug('Signing build assertion: {}'.format(cmd))
+                snapcraft.internal.common.run(
+                    cmd, stdout=outfile, stdin=subprocess.PIPE)
+            except subprocess.CalledProcessError:
+                msg = 'Failed to sign build assertion {}.'.format(assertion_file)
+                raise snapcraft.internal.meta.CommandError(msg)
+        logger.info('Assertion {} saved to disk.'.format(assertion_file))
+
+    def push_snap_build(self, account_info, snap_name, snap_filename):
+        assertion_file = sub(r"(.*).snap$", r"\1.snap-build", snap_filename)
+        try:
+            snaps = account_info['snaps'][constants.DEFAULT_SERIES]
+            snap_id = snaps[snap_name]['snap-id']
+        except KeyError as err:
+            msg = 'Could not get all account information needed'
+            raise snapcraft.internal.meta.CommandError(msg)
```

Comment:
This block is repeated. I'd make it `push_snap_build(self, snap_id, assertion)` and extract the inputs on the callsite.